frontend/src/components/problem/ProblemSelector.tsx

```diff
@@ -0,0 +1,144 @@
+import React, { useEffect, useRef, useState } from 'react';
+import styled from 'styled-components';
+import { getProblems, SelectableProblem } from '../../api/Problem';
+import ErrorMessage from '../core/Error';
+import { displayNameFromDifficulty } from '../../api/Difficulty';
+import { InlineDifficultyDisplayButton } from '../core/Button';
+import { TextInput } from '../core/Input';
+
+type ProblemSelectorProps = {
+  selectedProblems: SelectableProblem[],
+  onSelect: (newlySelected: SelectableProblem) => void,
+};
+
+type ContentProps = {
+  show: boolean,
+};
+
+const Content = styled.div`
+  width: 65%;
+  min-width: 250px;
+  margin: 10px 0;
+`;
+
+const InnerContent = styled.div<ContentProps>`
+  width: 100%;
+  max-height: 200px;
+  overflow-y: scroll;
+  display: ${({ show }) => (show ? 'block' : 'none')};
+  border-radius: 5px;
+  //border: solid 1px ${({ theme }) => theme.colors.border};
+  box-shadow: 0 1px 4px rgba(0, 0, 0, 0.24);
+`;
+
+const InlineProblem = styled.div`
+  width: 100%;
+  padding: 8px 15px;
+  display: flex;
+  flex: auto;
+  justify-content: space-between;
+  box-sizing: border-box;
+  border-bottom: solid 1px ${({ theme }) => theme.colors.background};
+
+  &:hover {
+    cursor: pointer;
+    background-color: ${({ theme }) => theme.colors.background};
+  }
+`;
+
+const ProblemSearch = styled(TextInput)`
+  width: 100%;
+  margin: 0;
+`;
+
+const ProblemName = styled.p`
+  font-weight: bold;
+  margin: 0;
+`;
+
+function ProblemSelector(props: ProblemSelectorProps) {
+  const { selectedProblems, onSelect } = props;
+
+  const [error, setError] = useState('');
+  const [problems, setProblems] = useState<SelectableProblem[]>([]);
+  const [showProblems, setShowProblems] = useState(false);
+  const [searchText, setSearchText] = useState('');
+
+  const ref = useRef<HTMLDivElement>(null);
+
+  // Close list of problems if clicked outside of div
+  useEffect(() => {
+    const handleClickOutside = (e: MouseEvent) => {
+      if (ref.current && !ref.current!.contains(e.target as Node)) {
+        setShowProblems(false);
+      }
+    };
+
+    document.addEventListener('mousedown', handleClickOutside);
+    return () => document.removeEventListener('mousedown', handleClickOutside);
+  }, [ref]);
```

Comment:
that's actually such a good point lol - essentially useRef here gives you access to the specified DOM node (at the `ref={ref}`via `ref.current`. then, it adds an event listener for mousedown that checks if it clicked inside of the chosen node (if not, then it will close the window)

Reply:
Do you think we should revise the current "Hover Element" setup we have with **ref** then when we revise the frontend?